src/functions/dialog/index.ts

```diff
@@ -5,38 +5,59 @@
 import type { Component } from 'vue'
 import { createApp, toRaw } from 'vue'
 
-interface DialogProps {
-	[index: string]: unknown
-	container?: string
+type SpawnDialogOptions = {
+	/**
+	 * Container to mount the dialog to
+	 * @default document.body
+	 */
+	container?: Element | string
 }
 
 /**
- * Helper to spawn a Vue dialog without having to mount it from a component
+ * Spawn a single-use Vue dialog instance to get the result when it is closed
  *
- * @param dialog The dialog component to spawn
- * @param props Properties to pass to the dialog
- * @param props.container Optionally pass a query selector for the dialog container element
- * @param onClose Callback when the dialog is closed
+ * @param dialog - Dialog component to spawn
+ * @param props - Props to pass to the dialog instance
+ * @param options - Spawning options or a callback when the dialog is closed
+ * @return Promise resolved with the `close` event payload
  */
 export function spawnDialog(
 	dialog: Component,
-	props?: DialogProps,
-	onClose: (...rest: unknown[]) => void = () => {},
-): void {
-	const el = document.createElement('div')
-	const container: HTMLElement = typeof props?.container === 'string'
-		? (document.querySelector(props.container) || document.body)
-		: document.body
-	container.appendChild(el)
+	props: object = {},
+	options: SpawnDialogOptions = {},
+): Promise<unknown> {
+	let { container } = options
 
-	const app = createApp(dialog, {
-		...props,
-		onClose: (...rest: unknown[]) => {
-			onClose(...rest.map(v => toRaw(v)))
-			app.unmount()
-			el.remove()
-		},
-	})
+	// For backwards compatibility try to use container from props
+	if ('container' in props && typeof props.container === 'string') {
+		container = props.container
+	}
+
+	// Resolve container to an Element or fallback to document.body
+	const resolvedContainer = (typeof container === 'string' ? document.querySelector(container) : container) || document.body
+
+	// Create root container element for the dialog
+	const element = resolvedContainer.appendChild(document.createElement('div'))
 
-	app.mount(el)
+	return new Promise((resolve, reject) => {
+		const app = createApp(dialog, {
+			...props,
+			// If dialog has no `container` prop passing a falsy value does nothing
+			// Otherwise it is expected that `null` disables teleport and mounts dialog in place like NcDialog/NcModal
+			container: null,
+			onClose(...rest: unknown[]) {
+				app.unmount()
+				element.remove()
+				const result = rest.map(v => toRaw(v))
+				resolve(result.length > 1 ? result : result[0])
+			},
+			'onVue:unmounted'() {
+				app.unmount()
+				element.remove()
+				reject(new Error('Dialog was unmounted without close event'))
```

Comment:
Maybe it makes sense to have a custom error to distinguish this, like:

```ts
class DialogUnmounted extends Error {}
```

But not sure if needed

Reply:
I don't think so because this is an actual exception, not an application error thrown as an exception. Function users never need to distinguish this error to handle it differently. It only says that the function is used wrong.